core/services/relay/evm/exec_provider.go

```diff
@@ -0,0 +1,391 @@
+	usdcAttestationAPITimeoutSeconds       int
+	usdcAttestationAPIIntervalMilliseconds int
+	usdcSrcMsgTransmitterAddr              common.Address
+
+	// these values are nil and are updated for Close()
+	seenOnRampAddress       *cciptypes.Address
+	seenSourceChainSelector *uint64
+	seenDestChainSelector   *uint64
+}
+
+func NewSrcExecProvider(
+	lggr logger.Logger,
+	versionFinder ccip.VersionFinder,
+	client client.Client,
+	estimator gas.EvmFeeEstimator,
+	maxGasPrice *big.Int,
+	lp logpoller.LogPoller,
+	startBlock uint64,
+	jobID string,
+	usdcAttestationAPI string,
+	usdcAttestationAPITimeoutSeconds int,
+	usdcAttestationAPIIntervalMilliseconds int,
+	usdcSrcMsgTransmitterAddr common.Address,
+) (commontypes.CCIPExecProvider, error) {
+	var usdcReader *ccip.USDCReaderImpl
+	var err error
+	if usdcAttestationAPI != "" {
+		usdcReader, err = ccip.NewUSDCReader(lggr, jobID, usdcSrcMsgTransmitterAddr, lp, true)
+		if err != nil {
+			return nil, fmt.Errorf("new usdc reader: %w", err)
+		}
+	}
+
+	return &SrcExecProvider{
+		lggr:                                   lggr,
+		versionFinder:                          versionFinder,
+		client:                                 client,
+		estimator:                              estimator,
+		maxGasPrice:                            maxGasPrice,
+		lp:                                     lp,
+		startBlock:                             startBlock,
+		usdcReader:                             usdcReader,
+		usdcAttestationAPI:                     usdcAttestationAPI,
+		usdcAttestationAPITimeoutSeconds:       usdcAttestationAPITimeoutSeconds,
+		usdcAttestationAPIIntervalMilliseconds: usdcAttestationAPIIntervalMilliseconds,
+		usdcSrcMsgTransmitterAddr:              usdcSrcMsgTransmitterAddr,
+	}, nil
+}
+
+func (s *SrcExecProvider) Name() string {
+	return "CCIP.SrcExecProvider"
+}
+
+func (s *SrcExecProvider) Start(ctx context.Context) error {
+	if s.startBlock != 0 {
+		s.lggr.Infow("start replaying src chain", "fromBlock", s.startBlock)
+		return s.lp.Replay(ctx, int64(s.startBlock))
+	}
+	return nil
+}
+
+// Close is called when the job that created this provider is closed.
+func (s *SrcExecProvider) Close() error {
+	versionFinder := ccip.NewEvmVersionFinder()
+
+	unregisterFuncs := make([]func() error, 0, 2)
+	unregisterFuncs = append(unregisterFuncs, func() error {
+		// avoid panic in the case NewOnRampReader wasn't called
+		if s.seenOnRampAddress == nil {
+			return nil
+		}
+		return ccip.CloseOnRampReader(s.lggr, versionFinder, *s.seenSourceChainSelector, *s.seenDestChainSelector, *s.seenOnRampAddress, s.lp, s.client)
+	})
+	unregisterFuncs = append(unregisterFuncs, func() error {
+		if s.usdcAttestationAPI == "" {
+			return nil
+		}
+		return ccip.CloseUSDCReader(s.lggr, s.lggr.Name(), s.usdcSrcMsgTransmitterAddr, s.lp)
+	})
+	var multiErr error
+	for _, fn := range unregisterFuncs {
+		if err := fn(); err != nil {
+			multiErr = multierr.Append(multiErr, err)
```

Comment:
We've been trying to drop this import in favor of the std lib:
```suggestion
			multiErr = errors.Join(multiErr, err)
```
That can be a follow up change though.